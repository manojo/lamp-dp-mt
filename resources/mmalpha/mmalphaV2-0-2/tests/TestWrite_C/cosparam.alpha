--From: Sophie Robert <Sophie.Robert@irisa.fr>

system cosparam (v1,v2: {i | 1<=i<=4} of real) returns (c,s:real);

	var 	alp: real;
		beta: real;
		gama: real;
		eps: real;
		t: real;
            	al: {j | 0<=j<=4} of real;
		be: {j | 0<=j<=4} of real;
		ga: {j | 0<=j<=4} of real;
let
--calcul de la norme de v1
	al =	case
		{j| j=0}: 0.(j->);
		{j| j>0}: al.(j->j-1) + v1.(j->j)*v1.(j->j);
		esac;
	alp = al.(->4);
--calcul de la norme de v2
	be =	case
		{j| j=0}: 0.(j->);
		{j| j>0}: be.(j->j-1) + v2.(j->j)*v2.(j->j);
		esac;
	beta = be.(->4);
--calcul du produit scalaire v1 et v2
	ga =	case
		{j| j=0}: 0.(j->);
		{j| j>0}: ga.(j->j-1) + v1.(j->j)*v2.(j->j);
		esac;
	gama = ga.(->4);

	eps = (beta - alp)/(2*gama);

	t = if eps=0 then 1 else if eps>0 then 1/(eps + (1+eps*eps)) else -1/(-eps + (1 + eps*eps));

	c = 1/(1 + t*t);
	
	s = c*t;

tel; 

