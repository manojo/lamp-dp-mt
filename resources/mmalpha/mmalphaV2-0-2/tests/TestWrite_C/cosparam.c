/* system cosparam */

#include <math.h>

typedef struct { int value;
		 int computed; } intvar;
typedef struct { int value;
		 int computed; } boolvar;
typedef struct { double value;
		 int computed; } realvar;
double atof();
int    atoi();

#define min(x,y) ((x)<(y)?(x):(y))
#define max(x,y) ((x)>(y)?(x):(y))
#define INFINITY 0x7fffffff

/* --parameters */

/* --input variables */

char _s[32];

realvar v1[4];
double compute_v1(i)
int i;
{
  realvar *tmp;
/*printf("Computing: v1[%d]\n" ,i);*/
  tmp = &(v1[i-1]);
  if (!tmp->computed) {
     printf("Input v1[%d] =" ,i);
     gets(_s);
     tmp->value = atof(_s);
     tmp->computed = 1;}
  return tmp->value;
}

realvar v2[4];
double compute_v2(i)
int i;
{
  realvar *tmp;
/*printf("Computing: v2[%d]\n" ,i);*/
  tmp = &(v2[i-1]);
  if (!tmp->computed) {
     printf("Input v2[%d] =" ,i);
     gets(_s);
     tmp->value = atof(_s);
     tmp->computed = 1;}
  return tmp->value;
}

/* --output variables */

realvar c;
double compute_c();

realvar s;
double compute_s();

/* --local variables */

realvar alp;
double compute_alp();

realvar beta;
double compute_beta();

realvar gama;
double compute_gama();

realvar eps;
double compute_eps();

realvar t;
double compute_t();

realvar al[5];
double compute_al();

realvar be[5];
double compute_be();

realvar ga[5];
double compute_ga();

/* --let equations */

double compute_al(j)
int j;
{
  realvar *tmp;
/*printf("Computing: al[%d]\n" ,j);*/
  tmp = &(al[j]);
  if (!tmp->computed) {
     tmp->value = ( j==0 )  ? ( 0 ) : 
         ( j-1>=0 )  ? ( compute_al(j-1) + compute_v1(j) * compute_v1(j) ) : 
         ( printf("? case error\n"), exit(-1));
     tmp->computed = 1;}
  return tmp->value;
}

double compute_alp()
{
  realvar *tmp;
/*printf("Computing: alp[]\n");*/
  tmp = &(alp);
  if (!tmp->computed) {
     tmp->value = compute_al(4);
     tmp->computed = 1;}
  return tmp->value;
}

double compute_be(j)
int j;
{
  realvar *tmp;
/*printf("Computing: be[%d]\n" ,j);*/
  tmp = &(be[j]);
  if (!tmp->computed) {
     tmp->value = ( j==0 )  ? ( 0 ) : 
         ( j-1>=0 )  ? ( compute_be(j-1) + compute_v2(j) * compute_v2(j) ) : 
         ( printf("? case error\n"), exit(-1));
     tmp->computed = 1;}
  return tmp->value;
}

double compute_beta()
{
  realvar *tmp;
/*printf("Computing: beta[]\n");*/
  tmp = &(beta);
  if (!tmp->computed) {
     tmp->value = compute_be(4);
     tmp->computed = 1;}
  return tmp->value;
}

double compute_ga(j)
int j;
{
  realvar *tmp;
/*printf("Computing: ga[%d]\n" ,j);*/
  tmp = &(ga[j]);
  if (!tmp->computed) {
     tmp->value = ( j==0 )  ? ( 0 ) : 
         ( j-1>=0 )  ? ( compute_ga(j-1) + compute_v1(j) * compute_v2(j) ) : 
         ( printf("? case error\n"), exit(-1));
     tmp->computed = 1;}
  return tmp->value;
}

double compute_gama()
{
  realvar *tmp;
/*printf("Computing: gama[]\n");*/
  tmp = &(gama);
  if (!tmp->computed) {
     tmp->value = compute_ga(4);
     tmp->computed = 1;}
  return tmp->value;
}

double compute_eps()
{
  realvar *tmp;
/*printf("Computing: eps[]\n");*/
  tmp = &(eps);
  if (!tmp->computed) {
     
     tmp->value = (compute_beta() - compute_alp()) / 2 * compute_gama();
     tmp->computed = 1;}
  return tmp->value;
}

double compute_t()
{
  realvar *tmp;
/*printf("Computing: t[]\n");*/
  tmp = &(t);
  if (!tmp->computed) {
     tmp->value = (compute_eps() == 0) ? (1) : (
             (compute_eps() > 0) ? 
             (1 / (compute_eps() + 1 + compute_eps() * compute_eps())) : 
             (-1 / (-compute_eps() + 1 + compute_eps() * compute_eps())));
     tmp->computed = 1;}
  return tmp->value;
}

double compute_c()
{
  realvar *tmp;
/*printf("Computing: c[]\n");*/
  tmp = &(c);
  if (!tmp->computed) {
     tmp->value = 1 / (1 + compute_t() * compute_t());
     tmp->computed = 1;}
  return tmp->value;
}

double compute_s()
{
  realvar *tmp;
/*printf("Computing: s[]\n");*/
  tmp = &(s);
  if (!tmp->computed) {
     tmp->value = compute_c() * compute_t();
     tmp->computed = 1;}
  return tmp->value;
}

int main()
{
   printf("c = %f\n" , compute_c() );
   
   printf("s = %f\n" , compute_s() );
   
}
