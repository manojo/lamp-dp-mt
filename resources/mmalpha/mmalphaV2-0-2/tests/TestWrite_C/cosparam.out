1.000000
0.000011
